File: rpg-project-backend/application/models/_inventory.py
from application import db
import enum

class InventoryType(enum.Enum):
    EQUIPED = "Equipped"
    CARRIED = "Carried"
    TRANSPORT = "Transport"

class Inventory(db.Model):
    __tablename__ = "inventory"

    id: int = db.Column(db.Integer, primary_key=True)
    character_id: int = db.Column(db.Integer, db.ForeignKey('character.id'), nullable=False)
    name: str = db.Column(db.String(80), nullable=False)
    description: str = db.Column(db.Text, nullable=False)
    inv_type: InventoryType = db.Column(db.Enum(InventoryType), nullable=False)
    capacity: float = db.Column(db.Integer, nullable=False, default=0)
    items = db.relationship('InventoryItem', backref='inventory', lazy=True, cascade='all, delete-orphan')


    def toDict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'type': self.inv_type.value,
            'capacity': self.capacity,
        }
    
class InventoryItem(db.Model):
    __tablename__ = "inventory_item"

    id: int = db.Column(db.Integer, primary_key=True)
    inventory_id: int = db.Column(db.Integer, db.ForeignKey('inventory.id'), nullable=False)
    item_id: int = db.Column(db.Integer, db.ForeignKey('item.id'), nullable=False)
    quantity: int = db.Column(db.Integer, nullable=False, default=1)

    def toDict(self):
        return {
            'id': self.id,
            'inventory_id': self.inventory_id,
            'item_id': self.item_id,
            'quantity': self.quantity
        }

class Item(db.Model):
    __tablename__ = "item"

    id: int = db.Column(db.Integer, primary_key=True)

    name: str = db.Column(db.String(80), nullable=False)
    description: str = db.Column(db.Text, nullable=False)
    stackable: bool = db.Column(db.Boolean, nullable=False, default=False)
    equipable: bool = db.Column(db.Boolean, nullable=False, default=False)
    item_type: str = db.Column(db.String(50), nullable=False)
    max_quantity: int = db.Column(db.Integer, nullable=True, default=1)
    temporary: bool = db.Column(db.Boolean, nullable=False, default=False)
    hidden: bool = db.Column(db.Boolean, nullable=False, default=True)

    def toDict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'stackable': self.stackable,
            'equipable': self.equipable,
            'max_quantity': self.max_quantity,
            'item_type': self.item_type,
            'hidden': self.hidden,
        }

    __mapper_args__ = {
        "polymorphic_on": item_type,
        "polymorphic_identity": "item",
    }
    
class Weapon(Item):
    id: int = db.Column(db.Integer, db.ForeignKey('item.id'), primary_key=True)

    damage: str = db.Column(db.String(100), nullable=False)
    pericia: str = db.Column(db.String(100), nullable=False)
    critical: str = db.Column(db.String(100), nullable=False)
    range: str = db.Column(db.String(100), nullable=False)
    
    def toDict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'stackable': self.stackable,
            'equipable': self.equipable,
            'damage': self.damage,
            'pericia': self.pericia,
            'critical': self.critical,
            'range': self.range,
            'max_quantity': self.max_quantity,
            'item_type': self.item_type,
            'temporary': self.temporary,
            'hidden': self.hidden
        }

    __mapper_args__ = {
        "polymorphic_identity": "weapon",
    }

class Armor(Item):
    id: int = db.Column(db.Integer, db.ForeignKey('item.id'), primary_key=True)

    resistance: str = db.Column(db.String(100), nullable=False)
    reduction: str = db.Column(db.String(100), nullable=False)
    pericia: str = db.Column(db.String(100), nullable=False)
    size: str = db.Column(db.String(100), nullable=False)
    effect: str = db.Column(db.String(100), nullable=True)

    def toDict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'stackable': self.stackable,
            'equipable': self.equipable,
            'resistance': self.resistance,
            'reduction': self.reduction,
            'pericia': self.pericia,
            'size': self.size,
            'max_quantity': self.max_quantity,
            'item_type': self.item_type,
            'temporary': self.temporary,
            'effect': self.effect,
            'hidden': self.hidden
        }

    __mapper_args__ = {
        "polymorphic_identity": "armor",
    }

class Artefact(Item):
    id: int = db.Column(db.Integer, db.ForeignKey('item.id'), primary_key=True)

    effect: str = db.Column(db.String(100), nullable=False)

    __mapper_args__ = {
        "polymorphic_identity": "artefact",
    }

    def toDict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'stackable': self.stackable,
            'equipable': self.equipable,
            'effect': self.effect,
            'max_quantity': self.max_quantity,
            'item_type': self.item_type,
            'temporary': self.temporary,
            'hidden': self.hidden
        }

class Utility(Item):
    id: int = db.Column(db.Integer, db.ForeignKey('item.id'), primary_key=True)

    __mapper_args__ = {
        "polymorphic_identity": "utility",
    }

    def toDict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'stackable': self.stackable,
            'equipable': self.equipable,
            'max_quantity': self.max_quantity,
            'item_type': self.item_type,
            'temporary': self.temporary,
            'hidden': self.hidden
        }

File: rpg-project-frontend/app/types/inventory.ts
// ==================== INVENTORIES ====================

export type InventoryType =
  | "CARRIED"
  | "EQUIPED"
  | "TRANSPORT"
  | string;

export type InventoryItem = {
  id: number;
  

  name: string;
  description: string;

  type: InventoryType;
  capacity: number;
};

export type ListInventoriesResponse = {
  inventories: InventoryItem[];
};

export type CreateInventoryRequest = {
  name: string;
  description: string;
  type: InventoryType;
  capacity?: number;
};

export type CreateInventoryResponse = {
  message: string;
  inventory: InventoryItem;
};

export type UpdateInventoryRequest = {
  name?: string;
  description?: string;
  capacity?: number;
};

export type UpdateInventoryResponse = {
  message: string;
  inventory: InventoryItem;
};

export type CreateStandardInventoriesResponse = {
  message: string;
  inventories: InventoryItem[];
};

export type InventoryTypesResponse = {
  inventory_types: InventoryType[];
};

export type TransferInventoryOwnershipResponse = {
  message: string;
  inventory: InventoryItem;
};

// ==================== ITEMS ====================

export type BaseItem = {
  id: number;

  name: string;
  description: string;

  stackable: boolean;
  equipable: boolean;
  max_quantity: number | null;
  temporary?: boolean;
  hidden?: boolean;

  item_type: "weapon" | "armor" | "artefact" | "utility";
};

export type ToggleItemVisibilityResponse = {
  message: string;
  item: Item;
};

export type ToggleItemTemporaryResponse = {
  message: string;
  item: Item;
};

export type WeaponItem = BaseItem & {
  item_type: "weapon";

  damage: string;
  pericia: string;
  critical: string;
  range: string;
};

export type ArmorItem = BaseItem & {
  item_type: "armor";

  resistance: string;
  reduction: string;
  pericia: string;
  size: string;
  effect: string;
};

export type ArtefactItem = BaseItem & {
  item_type: "artefact";

  effect: string;
};

export type UtilityItem = BaseItem & {
  item_type: "utility";
};

export type Item =
  | WeaponItem
  | ArmorItem
  | ArtefactItem
  | UtilityItem;

export type ListItemsResponse = {
  items: Item[];
};

export type GetItemResponse = {
  item: Item;
};

export type CreateItemRequest = {
  name: string;
  description: string;

  item_type: "weapon" | "armor" | "artefact" | "utility";

  stackable?: boolean;
  equipable?: boolean;
  max_quantity?: number;

  data?: Record<string, any>;
};

export type CreateItemResponse = {
  message: string;
  item: Item;
};

export type UpdateItemRequest = {
  name?: string;
  description?: string;

  stackable?: boolean;
  equipable?: boolean;
  max_quantity?: number;

  data?: Record<string, any>;
};

export type UpdateItemResponse = {
  message: string;
  item: Item;
};

// ==================== INVENTORY ITEMS ====================

export type InventoryEntry = Item & {
  inventory_id: number;
  item_id: number;
  quantity: number;
};

export type InventoryItemsResponse = {
  items: Record<string, InventoryEntry[]>;
};

export type AddInventoryItemRequest = {
  item_id: number;
  quantity?: number;
};

export type AddInventoryItemResponse = {
  message: string;
  inventory_item: InventoryEntry;
};

export type RemoveInventoryItemRequest = {
  quantity?: number;
};

export type RemoveInventoryItemResponse = {
  message: string;
  inventory_item: InventoryEntry | null;
};

export type TransferInventoryItemRequest = {
  item_id: number;
  quantity?: number;
};